// payload/dist/admin/components/forms/Form/initContextState.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const submit = ()=>undefined;
const getSiblingData = ()=>undefined;
const dispatchFields = ()=>undefined;
const validateForm = ()=>undefined;
const createFormData = ()=>undefined;
const setModified = ()=>undefined;
const setProcessing = ()=>undefined;
const setSubmitted = ()=>undefined;
const reset = ()=>undefined;
const initialContextState = {
    addFieldRow: ()=>undefined,
    buildRowErrors: ()=>undefined,
    createFormData,
    disabled: false,
    dispatchFields,
    fields: {},
    formRef: null,
    getData: ()=>undefined,
    getDataByPath: ()=>undefined,
    getField: ()=>undefined,
    getFields: ()=>({}),
    getSiblingData,
    removeFieldRow: ()=>undefined,
    replaceFieldRow: ()=>undefined,
    replaceState: ()=>undefined,
    reset,
    setModified,
    setProcessing,
    setSubmitted,
    submit,
    validateForm
};
const _default = initialContextState;
