// payload/dist/admin/components/elements/ReactSelect/MultiValue/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "MultiValue", {
    enumerable: true,
    get: function() {
        return MultiValue;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reactselect = require("react-select");
const _useDraggableSortable = require("../../DraggableSortable/useDraggableSortable");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'multi-value';
const MultiValue = (props)=>{
    const { className, data: { value }, innerProps, isDisabled, // @ts-expect-error // TODO Fix this - moduleResolution 16 breaks our declare module
    selectProps: { customProps: { disableMouseDown } = {}, isSortable } = {} } = props;
    const { attributes, isDragging, listeners, setNodeRef, transform } = (0, _useDraggableSortable.useDraggableSortable)({
        id: value.toString(),
        disabled: !isSortable
    });
    const classes = [
        baseClass,
        className,
        !isDisabled && 'draggable',
        isDragging && `${baseClass}--is-dragging`
    ].filter(Boolean).join(' ');
    return /*#__PURE__*/ _react.default.createElement(_reactselect.components.MultiValue, {
        ...props,
        className: classes,
        innerProps: {
            ...innerProps,
            ...attributes,
            ...listeners,
            onMouseDown: (e)=>{
                if (!disableMouseDown) {
                    // we need to prevent the dropdown from opening when clicking on the drag handle, but not when a modal is open (i.e. the 'Relationship' field component)
                    e.stopPropagation();
                }
            },
            ref: setNodeRef,
            style: {
                transform
            }
        }
    });
};
