// payload/dist/admin/components/views/Version/RenderFieldsToDiff/fields/Nested/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _ = /*#__PURE__*/ _interop_require_default(require("../.."));
const _getTranslation = require("../../../../../../../utilities/getTranslation");
const _Label = /*#__PURE__*/ _interop_require_default(require("../../Label"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'nested-diff';
const Nested = ({ comparison, disableGutter = false, field, fieldComponents, locale, locales, permissions, version })=>{
    const { i18n } = (0, _reacti18next.useTranslation)();
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: baseClass
    }, field.label && /*#__PURE__*/ _react.default.createElement(_Label.default, null, locale && /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__locale-label`
    }, locale), (0, _getTranslation.getTranslation)(field.label, i18n)), /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            `${baseClass}__wrap`,
            !disableGutter && `${baseClass}__wrap--gutter`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_.default, {
        comparison: comparison,
        fieldComponents: fieldComponents,
        fieldPermissions: permissions,
        fields: field.fields,
        locales: locales,
        version: version
    })));
};
const _default = Nested;
