// payload/dist/collections/graphql/resolvers/create.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return createResolver;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _create = /*#__PURE__*/ _interop_require_default(require("../../operations/create"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function createResolver(collection) {
    return async function resolver(_, args, context) {
        if (args.locale) {
            context.req.locale = args.locale;
        }
        const options = {
            collection,
            data: args.data,
            depth: 0,
            draft: args.draft,
            req: (0, _isolateTransactionID.default)(context.req)
        };
        const result = await (0, _create.default)(options);
        return result;
    };
}
