// payload/dist/collections/graphql/resolvers/findByID.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findByIDResolver;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _findByID = /*#__PURE__*/ _interop_require_default(require("../../operations/findByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findByIDResolver(collection) {
    return async function resolver(_, args, context) {
        const { req } = context;
        if (args.locale) req.locale = args.locale;
        if (args.fallbackLocale) req.fallbackLocale = args.fallbackLocale;
        const options = {
            id: args.id,
            collection,
            depth: 0,
            draft: args.draft,
            req: (0, _isolateTransactionID.default)(context.req)
        };
        const result = await (0, _findByID.default)(options);
        return result;
    };
}
