// payload/dist/collections/graphql/resolvers/findVersions.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findVersionsResolver;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _findVersions = /*#__PURE__*/ _interop_require_default(require("../../operations/findVersions"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findVersionsResolver(collection) {
    async function resolver(_, args, context) {
        if (args.locale) context.req.locale = args.locale;
        if (args.fallbackLocale) context.req.fallbackLocale = args.fallbackLocale;
        const options = {
            collection,
            depth: 0,
            limit: args.limit,
            page: args.page,
            req: (0, _isolateTransactionID.default)(context.req),
            sort: args.sort,
            where: args.where
        };
        const result = await (0, _findVersions.default)(options);
        return result;
    }
    return resolver;
}
