// payload/dist/collections/graphql/resolvers/find.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findResolver;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _find = /*#__PURE__*/ _interop_require_default(require("../../operations/find"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findResolver(collection) {
    return async function resolver(_, args, context) {
        if (args.locale) context.req.locale = args.locale;
        if (args.fallbackLocale) context.req.fallbackLocale = args.fallbackLocale;
        const options = {
            collection,
            depth: 0,
            draft: args.draft,
            limit: args.limit,
            page: args.page,
            req: (0, _isolateTransactionID.default)(context.req),
            sort: args.sort,
            where: args.where
        };
        const results = await (0, _find.default)(options);
        return results;
    };
}
