// payload/dist/auth/getExecuteStaticAccess.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _errors = require("../errors");
const _executeAccess = /*#__PURE__*/ _interop_require_default(require("./executeAccess"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const getExecuteStaticAccess = (config)=>async (req, res, next)=>{
        if (req.method === 'OPTIONS') {
            return res.sendStatus(200);
        }
        try {
            if (req.path) {
                const accessResult = await (0, _executeAccess.default)({
                    isReadingStaticFile: true,
                    req
                }, config.access.read);
                if (typeof accessResult === 'object') {
                    const filename = decodeURI(req.path).replace(/^\/|\/$/g, '');
                    const queryToBuild = {
                        and: [
                            {
                                or: [
                                    {
                                        filename: {
                                            equals: filename
                                        }
                                    }
                                ]
                            },
                            accessResult
                        ]
                    };
                    if (config.upload.imageSizes) {
                        config.upload.imageSizes.forEach(({ name })=>{
                            queryToBuild.and[0].or.push({
                                [`sizes.${name}.filename`]: {
                                    equals: filename
                                }
                            });
                        });
                    }
                    const doc = await req.payload.db.findOne({
                        collection: config.slug,
                        req,
                        where: queryToBuild
                    });
                    if (!doc) {
                        throw new _errors.Forbidden(req.t);
                    }
                }
            }
            return next();
        } catch (error) {
            return next(error);
        }
    };
const _default = getExecuteStaticAccess;
