// payload/dist/auth/baseFields/apiKey.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _crypto = /*#__PURE__*/ _interop_require_default(require("crypto"));
const _extractTranslations = require("../../translations/extractTranslations");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const labels = (0, _extractTranslations.extractTranslations)([
    'authentication:enableAPIKey',
    'authentication:apiKey'
]);
const encryptKey = ({ req, value })=>value ? req.payload.encrypt(value) : undefined;
const decryptKey = ({ req, value })=>value ? req.payload.decrypt(value) : undefined;
const _default = [
    {
        name: 'enableAPIKey',
        admin: {
            components: {
                Field: ()=>null
            }
        },
        defaultValue: false,
        label: labels['authentication:enableAPIKey'],
        type: 'checkbox'
    },
    {
        name: 'apiKey',
        admin: {
            components: {
                Field: ()=>null
            }
        },
        hooks: {
            afterRead: [
                decryptKey
            ],
            beforeChange: [
                encryptKey
            ]
        },
        label: labels['authentication:apiKey'],
        type: 'text'
    },
    {
        name: 'apiKeyIndex',
        admin: {
            disabled: true
        },
        hidden: true,
        hooks: {
            beforeValidate: [
                async ({ data, req, value })=>{
                    if (data.apiKey) {
                        return _crypto.default.createHmac('sha1', req.payload.secret).update(data.apiKey).digest('hex');
                    }
                    if (data.enableAPIKey === false) {
                        return null;
                    }
                    return value;
                }
            ]
        },
        type: 'text'
    }
];
