// payload/dist/utilities/configToJSONSchema.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
const _sanitize = require("../config/sanitize");
const _configToJSONSchema = require("./configToJSONSchema");
describe('configToJSONSchema', ()=>{
    it('should handle optional arrays with required fields', ()=>{
        const config = {
            collections: [
                {
                    fields: [
                        {
                            name: 'someRequiredField',
                            type: 'array',
                            fields: [
                                {
                                    name: 'someRequiredField',
                                    required: true,
                                    type: 'text'
                                }
                            ]
                        }
                    ],
                    slug: 'test',
                    timestamps: false
                }
            ]
        };
        const sanitizedConfig = (0, _sanitize.sanitizeConfig)(config);
        const schema = (0, _configToJSONSchema.configToJSONSchema)(sanitizedConfig, 'text');
        expect(schema?.definitions?.test).toStrictEqual({
            additionalProperties: false,
            properties: {
                id: {
                    type: 'string'
                },
                someRequiredField: {
                    items: {
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: [
                                    'string',
                                    'null'
                                ]
                            },
                            someRequiredField: {
                                type: 'string'
                            }
                        },
                        required: [
                            'someRequiredField'
                        ],
                        type: 'object'
                    },
                    type: [
                        'array',
                        'null'
                    ]
                }
            },
            required: [
                'id'
            ],
            title: 'Test',
            type: 'object'
        });
    });
});
