// payload/dist/admin/components/elements/Loading/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    LoadingOverlay: function() {
        return LoadingOverlay;
    },
    LoadingOverlayToggle: function() {
        return LoadingOverlayToggle;
    },
    FormLoadingOverlayToggle: function() {
        return FormLoadingOverlayToggle;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../utilities/getTranslation");
const _context = require("../../forms/Form/context");
const _LoadingOverlay = require("../../utilities/LoadingOverlay");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'loading-overlay';
const LoadingOverlay = ({ animationDuration, loadingText, overlayType, show = true })=>{
    const { t } = (0, _reacti18next.useTranslation)('general');
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            show ? `${baseClass}--entering` : `${baseClass}--exiting`,
            overlayType ? `${baseClass}--${overlayType}` : ''
        ].filter(Boolean).join(' '),
        style: {
            animationDuration: animationDuration || '500ms'
        }
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bars`
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bar`
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bar`
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bar`
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bar`
    }), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__bar`
    })), /*#__PURE__*/ _react.default.createElement("span", {
        className: `${baseClass}__text`
    }, loadingText || t('loading')));
};
const LoadingOverlayToggle = ({ name: key, loadingText, show, type = 'fullscreen' })=>{
    const { toggleLoadingOverlay } = (0, _LoadingOverlay.useLoadingOverlay)();
    _react.default.useEffect(()=>{
        toggleLoadingOverlay({
            isLoading: show,
            key,
            loadingText: loadingText || undefined,
            type
        });
        return ()=>{
            toggleLoadingOverlay({
                isLoading: false,
                key,
                type
            });
        };
    }, [
        show,
        toggleLoadingOverlay,
        key,
        type,
        loadingText
    ]);
    return null;
};
const FormLoadingOverlayToggle = ({ name, action, formIsLoading = false, loadingSuffix, type = 'fullscreen' })=>{
    const isProcessing = (0, _context.useFormProcessing)();
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const labels = {
        create: t('creating'),
        loading: t('loading'),
        update: t('updating')
    };
    return /*#__PURE__*/ _react.default.createElement(LoadingOverlayToggle, {
        loadingText: `${labels[action]} ${loadingSuffix ? (0, _getTranslation.getTranslation)(loadingSuffix, i18n) : ''}`.trim(),
        name: name,
        show: formIsLoading || isProcessing,
        type: type
    });
};
