// payload/dist/graphql/errorHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const errorHandler = async (payload, err, debug, afterErrorHook)=>{
    const status = err.originalError.status || _httpstatus.default.INTERNAL_SERVER_ERROR;
    let errorMessage = err.message;
    payload.logger.error(err.stack);
    // Internal server errors can contain anything, including potentially sensitive data.
    // Therefore, error details will be hidden from the response unless `config.debug` is `true`
    if (!debug && status === _httpstatus.default.INTERNAL_SERVER_ERROR) {
        errorMessage = 'Something went wrong.';
    }
    let response = {
        extensions: {
            name: err?.originalError?.name || undefined,
            data: err && err.originalError && err.originalError.data || undefined,
            stack: debug ? err.stack : undefined,
            statusCode: status
        },
        locations: err.locations,
        message: errorMessage,
        path: err.path
    };
    if (afterErrorHook) {
        ({ response } = await afterErrorHook(err, response, null, null) || {
            response
        });
    }
    return response;
};
const _default = errorHandler;
