// payload/dist/auth/strategies/local/register.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "registerLocalStrategy", {
    enumerable: true,
    get: function() {
        return registerLocalStrategy;
    }
});
const _errors = require("../../../errors");
const _generatePasswordSaltHash = require("./generatePasswordSaltHash");
const registerLocalStrategy = async ({ collection, doc, password, payload, req })=>{
    const existingUser = await payload.find({
        collection: collection.slug,
        depth: 0,
        where: {
            email: {
                equals: doc.email
            }
        }
    });
    if (existingUser.docs.length > 0) {
        throw new _errors.ValidationError([
            {
                field: 'email',
                message: 'A user with the given email is already registered'
            }
        ]);
    }
    const { hash, salt } = await (0, _generatePasswordSaltHash.generatePasswordSaltHash)({
        password
    });
    const sanitizedDoc = {
        ...doc
    };
    if (sanitizedDoc.password) delete sanitizedDoc.password;
    return payload.db.create({
        collection: collection.slug,
        data: {
            ...sanitizedDoc,
            hash,
            salt
        },
        req
    });
};
