// payload/dist/exports/config.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    buildConfig: function() {
        return _build.buildConfig;
    },
    defaults: function() {
        return _defaults.defaults;
    },
    sanitizeConfig: function() {
        return _sanitize.sanitizeConfig;
    },
    baseBlockFields: function() {
        return _baseBlockFields.baseBlockFields;
    },
    baseIDField: function() {
        return _baseIDField.baseIDField;
    },
    sanitizeFields: function() {
        return _sanitize1.sanitizeFields;
    }
});
const _build = require("../config/build");
_export_star(require("../config/types"), exports);
const _defaults = require("../config/defaults");
const _sanitize = require("../config/sanitize");
const _baseBlockFields = require("../fields/baseFields/baseBlockFields");
const _baseIDField = require("../fields/baseFields/baseIDField");
const _sanitize1 = require("../fields/config/sanitize");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
