// payload/dist/admin/components/graphics/Search/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const Search = ()=>/*#__PURE__*/ _react.default.createElement("svg", {
        className: "search",
        fill: "none",
        height: "26",
        viewBox: "0 0 25 26",
        width: "25",
        xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/ _react.default.createElement("circle", {
        className: "stroke",
        cx: "11.2069",
        cy: "10.9135",
        r: "5",
        stroke: "#333333",
        strokeWidth: "2"
    }), /*#__PURE__*/ _react.default.createElement("line", {
        className: "stroke",
        stroke: "#333333",
        strokeWidth: "2",
        x1: "14.914",
        x2: "20.5002",
        y1: "14.2063",
        y2: "19.7925"
    }));
const _default = Search;
