// payload/dist/admin/components/views/CreateFirstUser/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _RenderFields = /*#__PURE__*/ _interop_require_default(require("../../forms/RenderFields"));
const _Submit = /*#__PURE__*/ _interop_require_default(require("../../forms/Submit"));
const _fieldtypes = require("../../forms/field-types");
const _Minimal = /*#__PURE__*/ _interop_require_default(require("../../templates/Minimal"));
const _Auth = require("../../utilities/Auth");
const _Config = require("../../utilities/Config");
const _Meta = /*#__PURE__*/ _interop_require_default(require("../../utilities/Meta"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'create-first-user';
const CreateFirstUser = (props)=>{
    const { setInitialized } = props;
    const { fetchFullUser } = (0, _Auth.useAuth)();
    const { admin: { user: userSlug }, collections, routes: { admin, api }, serverURL } = (0, _Config.useConfig)();
    const { t } = (0, _reacti18next.useTranslation)('authentication');
    const userConfig = collections.find((collection)=>collection.slug === userSlug);
    const onSuccess = async (json)=>{
        if (json?.user?.token) {
            await fetchFullUser();
        }
        setInitialized(true);
    };
    const fields = [
        {
            name: 'email',
            label: t('general:emailAddress'),
            required: true,
            type: 'email'
        },
        {
            name: 'password',
            label: t('general:password'),
            required: true,
            type: 'password'
        },
        {
            name: 'confirm-password',
            label: t('confirmPassword'),
            required: true,
            type: 'confirmPassword'
        }
    ];
    return /*#__PURE__*/ _react.default.createElement(_Minimal.default, {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement("h1", null, t('general:welcome')), /*#__PURE__*/ _react.default.createElement("p", null, t('beginCreateFirstUser')), /*#__PURE__*/ _react.default.createElement(_Meta.default, {
        description: t('createFirstUser'),
        keywords: t('general:create'),
        title: t('createFirstUser')
    }), /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: `${serverURL}${api}/${userSlug}/first-register`,
        method: "post",
        onSuccess: onSuccess,
        redirect: admin,
        validationOperation: "create"
    }, /*#__PURE__*/ _react.default.createElement(_RenderFields.default, {
        fieldSchema: [
            ...fields,
            ...userConfig.fields
        ],
        fieldTypes: _fieldtypes.fieldTypes
    }), /*#__PURE__*/ _react.default.createElement(_Submit.default, null, t('general:create'))));
};
const _default = CreateFirstUser;
