// payload/dist/database/queryValidation/validateQueryPaths.js
/* eslint-disable no-restricted-syntax */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "validateQueryPaths", {
    enumerable: true,
    get: function() {
        return validateQueryPaths;
    }
});
const _QueryError = /*#__PURE__*/ _interop_require_default(require("../../errors/QueryError"));
const _constants = require("../../types/constants");
const _deepCopyObject = require("../../utilities/deepCopyObject");
const _flattenTopLevelFields = /*#__PURE__*/ _interop_require_default(require("../../utilities/flattenTopLevelFields"));
const _validateSearchParams = require("./validateSearchParams");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const flattenWhere = (query)=>Object.entries(query).reduce((flattenedConstraints, [key, val])=>{
        if ((key === 'and' || key === 'or') && Array.isArray(val)) {
            const subWhereConstraints = val.reduce((acc, subVal)=>{
                const subWhere = flattenWhere(subVal);
                return [
                    ...acc,
                    ...subWhere
                ];
            }, []);
            return [
                ...flattenedConstraints,
                ...subWhereConstraints
            ];
        }
        return [
            ...flattenedConstraints,
            {
                [key]: val
            }
        ];
    }, []);
async function validateQueryPaths({ collectionConfig, errors = [], globalConfig, overrideAccess, policies = {
    collections: {},
    globals: {}
}, req, versionFields, where }) {
    const fields = (0, _flattenTopLevelFields.default)(versionFields || (globalConfig || collectionConfig).fields);
    if (typeof where === 'object') {
        const whereFields = flattenWhere(where);
        // We need to determine if the whereKey is an AND, OR, or a schema path
        const promises = [];
        whereFields.map(async (constraint)=>{
            Object.keys(constraint).map(async (path)=>{
                Object.entries(constraint[path]).map(async ([operator, val])=>{
                    if (_constants.validOperators.includes(operator)) {
                        promises.push((0, _validateSearchParams.validateSearchParam)({
                            collectionConfig: (0, _deepCopyObject.deepCopyObject)(collectionConfig),
                            errors,
                            fields: fields,
                            globalConfig: (0, _deepCopyObject.deepCopyObject)(globalConfig),
                            operator,
                            overrideAccess,
                            path,
                            policies,
                            req,
                            val,
                            versionFields
                        }));
                    }
                });
            });
        });
        await Promise.all(promises);
        if (errors.length > 0) {
            throw new _QueryError.default(errors);
        }
    }
}
