// @payloadcms/db-mongodb/dist/migrateFresh.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "migrateFresh", {
    enumerable: true,
    get: function() {
        return migrateFresh;
    }
});
const _database = require("payload/database");
const _prompts = /*#__PURE__*/ _interop_require_default(require("prompts"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function migrateFresh() {
    const { payload } = this;
    const { confirm: acceptWarning } = await (0, _prompts.default)({
        name: 'confirm',
        initial: false,
        message: `WARNING: This will drop your database and run all migrations. Are you sure you want to proceed?`,
        type: 'confirm'
    }, {
        onCancel: ()=>{
            process.exit(0);
        }
    });
    if (!acceptWarning) {
        process.exit(0);
    }
    payload.logger.info({
        msg: `Dropping database.`
    });
    await this.connection.dropDatabase();
    const migrationFiles = await (0, _database.readMigrationFiles)({
        payload
    });
    payload.logger.debug({
        msg: `Found ${migrationFiles.length} migration files.`
    });
    let transactionID;
    // Run all migrate up
    for (const migration of migrationFiles){
        payload.logger.info({
            msg: `Migrating: ${migration.name}`
        });
        try {
            const start = Date.now();
            transactionID = await this.beginTransaction();
            await migration.up({
                payload
            });
            await payload.create({
                collection: 'payload-migrations',
                data: {
                    name: migration.name,
                    batch: 1
                },
                req: {
                    transactionID
                }
            });
            await this.commitTransaction(transactionID);
            payload.logger.info({
                msg: `Migrated:  ${migration.name} (${Date.now() - start}ms)`
            });
        } catch (err) {
            await this.rollbackTransaction(transactionID);
            payload.logger.error({
                err,
                msg: `Error running migration ${migration.name}. Rolling back.`
            });
            throw err;
        }
    }
}
