// @payloadcms/db-mongodb/dist/updateGlobalVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "updateGlobalVersion", {
    enumerable: true,
    get: function() {
        return updateGlobalVersion;
    }
});
const _withSession = require("./withSession");
async function updateGlobalVersion({ id, global, locale, req = {}, versionData, where }) {
    const VersionModel = this.versions[global];
    const whereToUse = where || {
        id: {
            equals: id
        }
    };
    const options = {
        ...(0, _withSession.withSession)(this, req.transactionID),
        lean: true,
        new: true
    };
    const query = await VersionModel.buildQuery({
        locale,
        payload: this.payload,
        where: whereToUse
    });
    const doc = await VersionModel.findOneAndUpdate(query, versionData, options);
    const result = JSON.parse(JSON.stringify(doc));
    const verificationToken = doc._verificationToken;
    // custom id type reset
    result.id = result._id;
    if (verificationToken) {
        result._verificationToken = verificationToken;
    }
    return result;
}
