// @payloadcms/db-mongodb/dist/updateOne.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "updateOne", {
    enumerable: true,
    get: function() {
        return updateOne;
    }
});
const _errors = require("payload/errors");
const _utilities = require("payload/utilities");
const _sanitizeInternalFields = /*#__PURE__*/ _interop_require_default(require("./utilities/sanitizeInternalFields"));
const _withSession = require("./withSession");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const updateOne = async function updateOne({ id, collection, data, locale, req = {}, where: whereArg }) {
    const where = id ? {
        id: {
            equals: id
        }
    } : whereArg;
    const Model = this.collections[collection];
    const options = {
        ...(0, _withSession.withSession)(this, req.transactionID),
        lean: true,
        new: true
    };
    const query = await Model.buildQuery({
        locale,
        payload: this.payload,
        where
    });
    let result;
    try {
        result = await Model.findOneAndUpdate(query, data, options);
    } catch (error) {
        // Handle uniqueness error from MongoDB
        throw error.code === 11000 && error.keyValue ? new _errors.ValidationError([
            {
                field: Object.keys(error.keyValue)[0],
                message: 'Value must be unique'
            }
        ], req?.t ?? (0, _utilities.i18nInit)(this.payload.config.i18n).t) : error;
    }
    result = JSON.parse(JSON.stringify(result));
    result.id = result._id;
    result = (0, _sanitizeInternalFields.default)(result);
    return result;
};
