// payload/dist/auth/graphql/resolvers/forgotPassword.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _forgotPassword = /*#__PURE__*/ _interop_require_default(require("../../operations/forgotPassword"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function forgotPasswordResolver(collection) {
    async function resolver(_, args, context) {
        const options = {
            collection,
            data: {
                email: args.email
            },
            disableEmail: args.disableEmail,
            expiration: args.expiration,
            req: (0, _isolateTransactionID.default)(context.req)
        };
        await (0, _forgotPassword.default)(options);
        return true;
    }
    return resolver;
}
const _default = forgotPasswordResolver;
