// payload/dist/auth/graphql/resolvers/login.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _login = /*#__PURE__*/ _interop_require_default(require("../../operations/login"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function loginResolver(collection) {
    async function resolver(_, args, context) {
        const options = {
            collection,
            data: {
                email: args.email,
                password: args.password
            },
            depth: 0,
            req: (0, _isolateTransactionID.default)(context.req),
            res: context.res
        };
        const result = (0, _login.default)(options);
        return result;
    }
    return resolver;
}
const _default = loginResolver;
