// payload/dist/auth/graphql/resolvers/resetPassword.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _resetPassword = /*#__PURE__*/ _interop_require_default(require("../../operations/resetPassword"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function resetPasswordResolver(collection) {
    async function resolver(_, args, context) {
        if (args.locale) context.req.locale = args.locale;
        if (args.fallbackLocale) context.req.fallbackLocale = args.fallbackLocale;
        const options = {
            api: 'GraphQL',
            collection,
            data: args,
            depth: 0,
            req: (0, _isolateTransactionID.default)(context.req),
            res: context.res
        };
        const result = await (0, _resetPassword.default)(options);
        return result;
    }
    return resolver;
}
const _default = resetPasswordResolver;
