// payload/dist/admin/components/forms/RowLabel/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "RowLabel", {
    enumerable: true,
    get: function() {
        return RowLabel;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../utilities/getTranslation");
const _context = require("../Form/context");
const _types = require("./types");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'row-label';
const RowLabel = ({ className, ...rest })=>{
    return /*#__PURE__*/ _react.default.createElement("span", {
        className: [
            baseClass,
            className
        ].filter(Boolean).join(' '),
        style: {
            pointerEvents: 'none'
        }
    }, /*#__PURE__*/ _react.default.createElement(RowLabelContent, rest));
};
const RowLabelContent = (props)=>{
    const { label, path, rowNumber } = props;
    const { i18n } = (0, _reacti18next.useTranslation)();
    const { getDataByPath, getSiblingData } = (0, _context.useWatchForm)();
    const collapsibleData = getSiblingData(path);
    const arrayData = getDataByPath(path);
    const data = arrayData || collapsibleData;
    if ((0, _types.isComponent)(label)) {
        const Label = label;
        return /*#__PURE__*/ _react.default.createElement(Label, {
            data: data,
            index: rowNumber,
            path: path
        });
    }
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, typeof label === 'function' ? label({
        data,
        index: rowNumber,
        path
    }) : (0, _getTranslation.getTranslation)(label, i18n));
};
