// payload/dist/versions/buildCollectionFields.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "buildVersionCollectionFields", {
    enumerable: true,
    get: function() {
        return buildVersionCollectionFields;
    }
});
const buildVersionCollectionFields = (collection)=>{
    const fields = [
        {
            name: 'parent',
            index: true,
            relationTo: collection.slug,
            type: 'relationship'
        },
        {
            name: 'version',
            fields: collection.fields,
            type: 'group'
        },
        {
            name: 'createdAt',
            admin: {
                disabled: true
            },
            index: true,
            type: 'date'
        },
        {
            name: 'updatedAt',
            admin: {
                disabled: true
            },
            index: true,
            type: 'date'
        }
    ];
    if (collection?.versions?.drafts) {
        fields.push({
            name: 'latest',
            admin: {
                disabled: true
            },
            index: true,
            type: 'checkbox'
        });
    }
    if (collection?.versions?.drafts && collection?.versions?.drafts?.autosave) {
        fields.push({
            name: 'autosave',
            index: true,
            type: 'checkbox'
        });
    }
    return fields;
};
