// payload/dist/admin/components/views/Global/Routes/CustomComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    defaultGlobalViews: function() {
        return defaultGlobalViews;
    },
    CustomGlobalComponent: function() {
        return CustomGlobalComponent;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _API = require("../../API");
const _LivePreview = require("../../LivePreview");
const _Version = /*#__PURE__*/ _interop_require_default(require("../../Version/Version"));
const _Versions = /*#__PURE__*/ _interop_require_default(require("../../Versions"));
const _index = require("../Default/index");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const defaultGlobalViews = {
    API: _API.API,
    Default: _index.DefaultGlobalEdit,
    LivePreview: _LivePreview.LivePreviewView,
    References: null,
    Relationships: null,
    Version: _Version.default,
    Versions: _Versions.default
};
const CustomGlobalComponent = (args)=>{
    const { global, view } = args;
    const { admin: { components: { views: { Edit } = {} } = {} } = {} } = global;
    // Overriding components may come from multiple places in the config
    // Need to cascade through the hierarchy to find the correct component to render
    // For example, the Edit view:
    // 1. Edit?.Default
    // 2. Edit?.Default?.Component
    // TODO: Remove the `@ts-ignore` when a Typescript wizard arrives
    // For some reason `Component` does not exist on type `Edit[view]` no matter how narrow the type is
    const Component = typeof Edit === 'object' && typeof Edit[view] === 'function' ? Edit[view] : typeof Edit === 'object' && typeof Edit?.[view] === 'object' && // @ts-ignore
    typeof Edit[view].Component === 'function' ? Edit[view].Component : defaultGlobalViews[view];
    if (Component) {
        return /*#__PURE__*/ _react.default.createElement(Component, args);
    }
};
