// payload/dist/admin/components/views/Global/Default.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _getTranslation = require("../../../../utilities/getTranslation");
const _DocumentHeader = require("../../elements/DocumentHeader");
const _Loading = require("../../elements/Loading");
const _Form = /*#__PURE__*/ _interop_require_default(require("../../forms/Form"));
const _OperationProvider = require("../../utilities/OperationProvider");
const _SetStepNav = require("../collections/Edit/SetStepNav");
const _Routes = require("./Routes");
const _CustomComponent = require("./Routes/CustomComponent");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'global-edit';
const DefaultGlobalView = (props)=>{
    const { i18n } = (0, _reacti18next.useTranslation)('general');
    const { action, apiURL, data, disableRoutes, fieldTypes, global, initialState, isLoading, onSave, permissions } = props;
    const { label } = global;
    const hasSavePermission = permissions?.update?.permission;
    return /*#__PURE__*/ _react.default.createElement("main", {
        className: baseClass
    }, /*#__PURE__*/ _react.default.createElement(_OperationProvider.OperationContext.Provider, {
        value: "update"
    }, /*#__PURE__*/ _react.default.createElement(_SetStepNav.SetStepNav, {
        global: global
    }), /*#__PURE__*/ _react.default.createElement(_Form.default, {
        action: action,
        className: `${baseClass}__form`,
        disabled: !hasSavePermission,
        initialState: initialState,
        method: "post",
        onSuccess: onSave
    }, /*#__PURE__*/ _react.default.createElement(_Loading.FormLoadingOverlayToggle, {
        action: "update",
        loadingSuffix: (0, _getTranslation.getTranslation)(label, i18n),
        name: `global-edit--${typeof label === 'string' ? label : label?.en}`
    }), !isLoading && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_DocumentHeader.DocumentHeader, {
        apiURL: apiURL,
        data: data,
        global: global
    }), disableRoutes ? /*#__PURE__*/ _react.default.createElement(_CustomComponent.CustomGlobalComponent, {
        view: "Default",
        ...props
    }) : /*#__PURE__*/ _react.default.createElement(_Routes.GlobalRoutes, {
        ...props,
        fieldTypes: fieldTypes
    })))));
};
const _default = DefaultGlobalView;
