// @payloadcms/bundler-webpack/dist/configs/dev.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getDevConfig", {
    enumerable: true,
    get: function() {
        return getDevConfig;
    }
});
const _md5 = /*#__PURE__*/ _interop_require_default(require("md5"));
const _webpack = /*#__PURE__*/ _interop_require_default(require("webpack"));
const _base = require("./base");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const getDevConfig = (payloadConfig)=>{
    const baseConfig = (0, _base.getBaseConfig)(payloadConfig);
    let webpackConfig = {
        ...baseConfig,
        cache: {
            type: 'filesystem',
            // version cache when there are changes to aliases
            buildDependencies: {
                config: [
                    __filename
                ]
            },
            version: (0, _md5.default)(Object.entries(baseConfig.resolve.alias).join())
        },
        devtool: 'inline-source-map',
        entry: {
            ...baseConfig.entry,
            main: [
                `${require.resolve('webpack-hot-middleware/client')}?path=${payloadConfig.routes.admin}/__webpack_hmr`,
                ...baseConfig.entry.main
            ]
        },
        mode: 'development',
        module: {
            ...baseConfig.module,
            rules: [
                ...baseConfig.module.rules,
                {
                    sideEffects: true,
                    test: /\.(scss|css)$/,
                    /*
           * The loaders here are run in reverse order. Here is how your loaders are being processed:
           * 1. sass-loader: This loader compiles your SCSS into CSS.
           * 2. postcss-loader: This loader applies postcss transformations (with preset-env plugin in your case).
           * 3. css-loader: This loader interprets @import and url() like import/require() and will resolve them.
           * 4. style-loader: This loader injects CSS into the DOM.
           */ use: [
                        require.resolve('style-loader'),
                        {
                            loader: require.resolve('css-loader'),
                            options: {
                                url: (url)=>!url.startsWith('/')
                            }
                        },
                        {
                            loader: require.resolve('postcss-loader'),
                            options: {
                                postcssOptions: {
                                    plugins: [
                                        require.resolve('postcss-preset-env')
                                    ]
                                }
                            }
                        },
                        require.resolve('sass-loader')
                    ]
                }
            ]
        },
        output: {
            filename: '[name].js',
            path: '/',
            publicPath: `${payloadConfig.routes.admin}/`
        },
        plugins: [
            ...baseConfig.plugins,
            new _webpack.default.HotModuleReplacementPlugin()
        ],
        stats: 'errors-warnings'
    };
    if (payloadConfig.admin.webpack && typeof payloadConfig.admin.webpack === 'function') {
        webpackConfig = payloadConfig.admin.webpack(webpackConfig);
    }
    return webpackConfig;
};
