// @payloadcms/bundler-webpack/dist/configs/prod.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getProdConfig", {
    enumerable: true,
    get: function() {
        return getProdConfig;
    }
});
const _minicssextractplugin = /*#__PURE__*/ _interop_require_default(require("mini-css-extract-plugin"));
const _swcminifywebpackplugin = require("swc-minify-webpack-plugin");
const _webpackbundleanalyzer = require("webpack-bundle-analyzer");
const _base = require("./base");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const getProdConfig = (payloadConfig)=>{
    const baseConfig = (0, _base.getBaseConfig)(payloadConfig);
    let webpackConfig = {
        ...baseConfig,
        mode: 'production',
        module: {
            ...baseConfig.module,
            rules: [
                ...baseConfig.module.rules,
                {
                    sideEffects: true,
                    test: /\.(scss|css)$/,
                    use: [
                        _minicssextractplugin.default.loader,
                        {
                            loader: require.resolve('css-loader'),
                            options: {
                                url: (url)=>!url.startsWith('/')
                            }
                        },
                        {
                            loader: require.resolve('postcss-loader'),
                            options: {
                                postcssOptions: {
                                    plugins: [
                                        require.resolve('postcss-preset-env')
                                    ]
                                }
                            }
                        },
                        require.resolve('sass-loader')
                    ]
                }
            ]
        },
        optimization: {
            minimizer: [
                new _swcminifywebpackplugin.SwcMinifyWebpackPlugin()
            ],
            splitChunks: {
                cacheGroups: {
                    styles: {
                        name: 'styles',
                        chunks: 'all',
                        enforce: true,
                        test: /\.(sa|sc|c)ss$/
                    }
                }
            }
        },
        output: {
            chunkFilename: '[name].[chunkhash].js',
            filename: '[name].[chunkhash].js',
            path: payloadConfig.admin.buildPath,
            publicPath: `${payloadConfig.routes.admin}/`
        },
        plugins: [
            ...baseConfig.plugins,
            new _minicssextractplugin.default({
                filename: '[name].[contenthash].css',
                ignoreOrder: true
            }),
            ...process.env.PAYLOAD_ANALYZE_BUNDLE ? [
                new _webpackbundleanalyzer.BundleAnalyzerPlugin()
            ] : []
        ],
        stats: 'errors-only'
    };
    if (payloadConfig.admin.webpack && typeof payloadConfig.admin.webpack === 'function') {
        webpackConfig = payloadConfig.admin.webpack(webpackConfig);
    }
    return webpackConfig;
};
