// @payloadcms/richtext-slate/dist/field/elements/Button.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    baseClass: function() {
        return baseClass;
    },
    default: function() {
        return _default;
    }
});
const _components = require("payload/components");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _slatereact = require("slate-react");
require("../buttons.scss");
const _isActive = /*#__PURE__*/ _interop_require_default(require("./isActive"));
const _toggle = /*#__PURE__*/ _interop_require_default(require("./toggle"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'rich-text__button';
const ElementButton = (props)=>{
    const { children, className, el = 'button', format, onClick, tooltip, type = 'type' } = props;
    const editor = (0, _slatereact.useSlate)();
    const [showTooltip, setShowTooltip] = (0, _react.useState)(false);
    const defaultOnClick = (0, _react.useCallback)((event)=>{
        event.preventDefault();
        setShowTooltip(false);
        (0, _toggle.default)(editor, format, type);
    }, [
        editor,
        format,
        type
    ]);
    const Tag = el;
    return /*#__PURE__*/ _react.default.createElement(Tag, {
        ...el === 'button' && {
            type: 'button'
        },
        className: [
            baseClass,
            className,
            (0, _isActive.default)(editor, format, type) && `${baseClass}__button--active`
        ].filter(Boolean).join(' '),
        onClick: onClick || defaultOnClick,
        onMouseEnter: ()=>setShowTooltip(true),
        onMouseLeave: ()=>setShowTooltip(false)
    }, tooltip && /*#__PURE__*/ _react.default.createElement(_components.Tooltip, {
        show: showTooltip
    }, tooltip), children);
};
const _default = ElementButton;
