// @payloadcms/richtext-slate/dist/field/elements/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _blockquote = /*#__PURE__*/ _interop_require_default(require("./blockquote"));
const _h1 = /*#__PURE__*/ _interop_require_default(require("./h1"));
const _h2 = /*#__PURE__*/ _interop_require_default(require("./h2"));
const _h3 = /*#__PURE__*/ _interop_require_default(require("./h3"));
const _h4 = /*#__PURE__*/ _interop_require_default(require("./h4"));
const _h5 = /*#__PURE__*/ _interop_require_default(require("./h5"));
const _h6 = /*#__PURE__*/ _interop_require_default(require("./h6"));
const _indent = /*#__PURE__*/ _interop_require_default(require("./indent"));
const _li = /*#__PURE__*/ _interop_require_default(require("./li"));
const _link = /*#__PURE__*/ _interop_require_default(require("./link"));
const _ol = /*#__PURE__*/ _interop_require_default(require("./ol"));
const _relationship = /*#__PURE__*/ _interop_require_default(require("./relationship"));
const _textAlign = /*#__PURE__*/ _interop_require_default(require("./textAlign"));
const _ul = /*#__PURE__*/ _interop_require_default(require("./ul"));
const _upload = /*#__PURE__*/ _interop_require_default(require("./upload"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const elements = {
    blockquote: _blockquote.default,
    h1: _h1.default,
    h2: _h2.default,
    h3: _h3.default,
    h4: _h4.default,
    h5: _h5.default,
    h6: _h6.default,
    indent: _indent.default,
    li: _li.default,
    link: _link.default,
    ol: _ol.default,
    relationship: _relationship.default,
    textAlign: _textAlign.default,
    ul: _ul.default,
    upload: _upload.default
};
const _default = elements;
