// @payloadcms/richtext-slate/dist/field/elements/toggleList.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _slate = require("slate");
const _slatereact = require("slate-react");
const _getCommonBlock = require("./getCommonBlock");
const _isListActive = /*#__PURE__*/ _interop_require_default(require("./isListActive"));
const _listTypes = /*#__PURE__*/ _interop_require_default(require("./listTypes"));
const _unwrapList = require("./unwrapList");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const toggleList = (editor, format)=>{
    let currentListFormat;
    if ((0, _isListActive.default)(editor, 'ol')) currentListFormat = 'ol';
    if ((0, _isListActive.default)(editor, 'ul')) currentListFormat = 'ul';
    // If the format is currently active,
    // remove the list
    if (currentListFormat === format) {
        const selectedLeaf = _slate.Node.descendant(editor, editor.selection.anchor.path);
        // If on an empty bullet, leave the above list alone
        // and unwrap only the active bullet
        if (_slate.Text.isText(selectedLeaf) && String(selectedLeaf.text).length === 0) {
            _slate.Transforms.unwrapNodes(editor, {
                match: (n)=>_slate.Element.isElement(n) && _listTypes.default.includes(n.type),
                mode: 'lowest',
                split: true
            });
            _slate.Transforms.setNodes(editor, {
                type: undefined
            });
        } else {
            // Otherwise, we need to unset li on all lis in the parent list
            // and unwrap the parent list itself
            const [, listPath] = (0, _getCommonBlock.getCommonBlock)(editor, (n)=>_slate.Element.isElement(n) && n.type === format);
            (0, _unwrapList.unwrapList)(editor, listPath);
        }
    // Otherwise, if a list is active and we are changing it,
    // change it
    } else if (currentListFormat && currentListFormat !== format) {
        _slate.Transforms.setNodes(editor, {
            type: format
        }, {
            match: (node)=>_slate.Element.isElement(node) && _listTypes.default.includes(node.type),
            mode: 'lowest'
        });
    // Otherwise we can assume that we should just activate the list
    } else {
        _slate.Transforms.wrapNodes(editor, {
            children: [],
            type: format
        });
        const [, parentNodePath] = (0, _getCommonBlock.getCommonBlock)(editor, (node)=>_slate.Element.isElement(node) && node.type === format);
        // Only set li on nodes that don't have type
        _slate.Transforms.setNodes(editor, {
            type: 'li'
        }, {
            match: (node, path)=>{
                const match = _slate.Element.isElement(node) && typeof node.type === 'undefined' && path.length === parentNodePath.length + 1;
                return match;
            },
            voids: true
        });
        // Wrap nodes that do have a type with an li
        // so as to not lose their existing formatting
        const nodesToWrap = Array.from(_slate.Editor.nodes(editor, {
            match: (node, path)=>{
                const match = _slate.Element.isElement(node) && typeof node.type !== 'undefined' && node.type !== 'li' && path.length === parentNodePath.length + 1;
                return match;
            }
        }));
        nodesToWrap.forEach(([, path])=>{
            _slate.Transforms.wrapNodes(editor, {
                children: [],
                type: 'li'
            }, {
                at: path
            });
        });
    }
    _slatereact.ReactEditor.focus(editor);
};
const _default = toggleList;
