// payload/components/forms.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Error: function() {
        return _Error.default;
    },
    FieldDescription: function() {
        return _FieldDescription.default;
    },
    Form: function() {
        return _Form.default;
    },
    buildInitialState: function() {
        return _buildInitialState.default;
    },
    useAllFormFields: function() {
        return _context.useAllFormFields;
    },
    useForm: function() {
        return _context.useForm;
    },
    useFormFields: function() {
        return _context.useFormFields;
    },
    useFormModified: function() {
        return _context.useFormModified;
    },
    useFormProcessing: function() {
        return _context.useFormProcessing;
    },
    useFormSubmitted: function() {
        return _context.useFormSubmitted;
    },
    /**
   * @deprecated useWatchForm is no longer preferred. If you need all form fields, prefer `useAllFormFields`.
   */ useWatchForm: function() {
        return _context.useWatchForm;
    },
    createNestedFieldPath: function() {
        return _createNestedFieldPath.createNestedFieldPath;
    },
    getSiblingData: function() {
        return _getSiblingData.default;
    },
    reduceFieldsToValues: function() {
        return _reduceFieldsToValues.default;
    },
    Label: function() {
        return _Label.default;
    },
    RenderFields: function() {
        return _RenderFields.default;
    },
    Submit: function() {
        return _Submit.default;
    },
    FormSubmit: function() {
        return _Submit.default;
    },
    fieldTypes: function() {
        return _fieldtypes.fieldTypes;
    },
    Checkbox: function() {
        return _Checkbox.default;
    },
    Collapsible: function() {
        return _Collapsible.default;
    },
    Date: function() {
        return _DateTime.default;
    },
    DateTimeInput: function() {
        return _Input.DateTimeInput;
    },
    Group: function() {
        return _Group.default;
    },
    HiddenInput: function() {
        return _HiddenInput.default;
    },
    Select: function() {
        return _Select.default;
    },
    SelectInput: function() {
        return _Input1.default;
    },
    Text: function() {
        return _Text.default;
    },
    TextInput: function() {
        return _Input2.default;
    },
    useFieldType: function() {
        return _useField.default;
    },
    useField: function() {
        return _useField.default;
    },
    withCondition: function() {
        return _withCondition.default;
    }
});
const _Error = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Error"));
const _FieldDescription = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/FieldDescription"));
const _Form = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Form"));
const _buildInitialState = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Form/buildInitialState"));
const _context = require("../dist/admin/components/forms/Form/context");
const _createNestedFieldPath = require("../dist/admin/components/forms/Form/createNestedFieldPath");
const _getSiblingData = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Form/getSiblingData"));
const _reduceFieldsToValues = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Form/reduceFieldsToValues"));
const _Label = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Label"));
const _RenderFields = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/RenderFields"));
const _Submit = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/Submit"));
const _fieldtypes = require("../dist/admin/components/forms/field-types");
const _Checkbox = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Checkbox"));
const _Collapsible = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Collapsible"));
const _DateTime = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/DateTime"));
const _Input = require("../dist/admin/components/forms/field-types/DateTime/Input");
const _Group = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Group"));
const _HiddenInput = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/HiddenInput"));
const _Select = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Select"));
const _Input1 = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Select/Input"));
const _Text = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Text"));
const _Input2 = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/field-types/Text/Input"));
const _useField = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/useField"));
const _withCondition = /*#__PURE__*/ _interop_require_default(require("../dist/admin/components/forms/withCondition"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
