// payload/dist/admin/components/elements/Nav/NavToggler/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "NavToggler", {
    enumerable: true,
    get: function() {
        return NavToggler;
    }
});
const _windowinfo = require("@faceless-ui/window-info");
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _Preferences = require("../../../utilities/Preferences");
const _context = require("../context");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'nav-toggler';
const NavToggler = (props)=>{
    const { id, children, className, tabIndex = 0 } = props;
    const { t } = (0, _reacti18next.useTranslation)('general');
    const { setPreference } = (0, _Preferences.usePreferences)();
    const { navOpen, setNavOpen } = (0, _context.useNav)();
    const { breakpoints: { l: largeBreak } } = (0, _windowinfo.useWindowInfo)();
    return /*#__PURE__*/ _react.default.createElement("button", {
        "aria-label": `${navOpen ? t('close') : t('open')} ${t('menu')}`,
        className: [
            baseClass,
            navOpen && `${baseClass}--is-open`,
            className
        ].filter(Boolean).join(' '),
        id: id,
        onClick: async ()=>{
            setNavOpen(!navOpen);
            // only when the user explicitly toggles the nav on desktop do we want to set the preference
            // this is because the js may open or close the nav based on the window size, routing, etc
            if (!largeBreak) {
                await setPreference('nav', {
                    open: !navOpen
                });
            }
        },
        tabIndex: tabIndex,
        type: "button"
    }, children);
};
