// payload/dist/collections/requestHandlers/find.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, // eslint-disable-next-line @typescript-eslint/no-explicit-any
"default", {
    enumerable: true,
    get: function() {
        return findHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _isNumber = require("../../utilities/isNumber");
const _find = /*#__PURE__*/ _interop_require_default(require("../operations/find"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function findHandler(req, res, next) {
    try {
        let page;
        if (typeof req.query.page === 'string') {
            const parsedPage = parseInt(req.query.page, 10);
            if (!Number.isNaN(parsedPage)) {
                page = parsedPage;
            }
        }
        const result = await (0, _find.default)({
            collection: req.collection,
            depth: (0, _isNumber.isNumber)(req.query.depth) ? Number(req.query.depth) : undefined,
            draft: req.query.draft === 'true',
            limit: (0, _isNumber.isNumber)(req.query.limit) ? Number(req.query.limit) : undefined,
            page,
            req,
            sort: req.query.sort,
            where: req.query.where
        });
        return res.status(_httpstatus.default.OK).json(result);
    } catch (error) {
        return next(error);
    }
}
