// payload/dist/collections/requestHandlers/updateByID.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    deprecatedUpdate: function() {
        return deprecatedUpdate;
    },
    default: function() {
        return updateByIDHandler;
    }
});
const _httpstatus = /*#__PURE__*/ _interop_require_default(require("http-status"));
const _formatSuccess = /*#__PURE__*/ _interop_require_default(require("../../express/responses/formatSuccess"));
const _updateByID = /*#__PURE__*/ _interop_require_default(require("../operations/updateByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function deprecatedUpdate(req, res, next) {
    req.payload.logger.warn('The PUT method is deprecated and will no longer be supported in a future release. Please use the PATCH method for update requests.');
    return updateByIDHandler(req, res, next);
}
async function updateByIDHandler(req, res, next) {
    try {
        const draft = req.query.draft === 'true';
        const autosave = req.query.autosave === 'true';
        const doc = await (0, _updateByID.default)({
            id: req.params.id,
            autosave,
            collection: req.collection,
            data: req.body,
            depth: parseInt(String(req.query.depth), 10),
            draft,
            req
        });
        let message = req.t('general:updatedSuccessfully');
        if (draft) message = req.t('version:draftSavedSuccessfully');
        if (autosave) message = req.t('version:autosavedSuccessfully');
        res.status(_httpstatus.default.OK).json({
            ...(0, _formatSuccess.default)(message, 'message'),
            doc
        });
    } catch (error) {
        next(error);
    }
}
