// payload/dist/globals/graphql/resolvers/findVersionByID.js
/* eslint-disable no-param-reassign */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return findVersionByIDResolver;
    }
});
const _isolateTransactionID = /*#__PURE__*/ _interop_require_default(require("../../../utilities/isolateTransactionID"));
const _findVersionByID = /*#__PURE__*/ _interop_require_default(require("../../operations/findVersionByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function findVersionByIDResolver(globalConfig) {
    return async function resolver(_, args, context) {
        if (args.locale) context.req.locale = args.locale;
        if (args.fallbackLocale) context.req.fallbackLocale = args.fallbackLocale;
        const options = {
            id: args.id,
            depth: 0,
            draft: args.draft,
            globalConfig,
            req: (0, _isolateTransactionID.default)(context.req)
        };
        const result = await (0, _findVersionByID.default)(options);
        return result;
    };
}
