// payload/dist/admin/components/elements/ThumbnailCard/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ThumbnailCard", {
    enumerable: true,
    get: function() {
        return ThumbnailCard;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _useTitle = require("../../../hooks/useTitle");
const _Config = require("../../utilities/Config");
const _Thumbnail = /*#__PURE__*/ _interop_require_default(require("../Thumbnail"));
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const baseClass = 'thumbnail-card';
const ThumbnailCard = (props)=>{
    const { alignLabel, className, collection, doc, label: labelFromProps, onClick, thumbnail } = props;
    const { i18n, t } = (0, _reacti18next.useTranslation)('general');
    const config = (0, _Config.useConfig)();
    const classes = [
        baseClass,
        className,
        typeof onClick === 'function' && `${baseClass}--has-on-click`,
        alignLabel && `${baseClass}--align-label-${alignLabel}`
    ].filter(Boolean).join(' ');
    let title = labelFromProps;
    if (!title) {
        title = (0, _useTitle.formatUseAsTitle)({
            collection,
            config,
            doc,
            i18n
        }) || (doc?.filename) || `[${t('untitled')}]`;
    }
    return /*#__PURE__*/ _react.default.createElement("button", {
        className: classes,
        onClick: onClick,
        title: title,
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__thumbnail`
    }, thumbnail && thumbnail, !thumbnail && collection && doc && /*#__PURE__*/ _react.default.createElement(_Thumbnail.default, {
        collection: collection,
        doc: doc,
        size: "expand"
    })), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__label`
    }, title));
};
