// @payloadcms/richtext-slate/dist/field/elements/link/LinkDrawer/index.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "LinkDrawer", {
    enumerable: true,
    get: function() {
        return LinkDrawer;
    }
});
const _elements = require("payload/components/elements");
const _forms = require("payload/components/forms");
const _hooks = require("payload/components/hooks");
const _utilities = require("payload/components/utilities");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
require("./index.scss");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'rich-text-link-edit-modal';
const LinkDrawer = ({ drawerSlug, fieldSchema, handleModalSubmit, initialState })=>{
    const { t } = (0, _reacti18next.useTranslation)('fields');
    return /*#__PURE__*/ _react.default.createElement(_elements.Drawer, {
        className: baseClass,
        slug: drawerSlug,
        title: t('editLink')
    }, /*#__PURE__*/ _react.default.createElement(_forms.Form, {
        fields: fieldSchema,
        initialState: initialState,
        onSubmit: handleModalSubmit
    }, /*#__PURE__*/ _react.default.createElement(_forms.RenderFields, {
        fieldSchema: fieldSchema,
        fieldTypes: _forms.fieldTypes,
        forceRender: true,
        readOnly: false
    }), /*#__PURE__*/ _react.default.createElement(LinkSubmit, null)));
};
const LinkSubmit = ()=>{
    const { t } = (0, _reacti18next.useTranslation)('fields');
    const ref = (0, _react.useRef)(null);
    const editDepth = (0, _utilities.useEditDepth)();
    (0, _hooks.useHotkey)({
        cmdCtrlKey: true,
        editDepth,
        keyCodes: [
            's'
        ]
    }, (e)=>{
        e.preventDefault();
        e.stopPropagation();
        if (ref?.current) {
            ref.current.click();
        }
    });
    return /*#__PURE__*/ _react.default.createElement(_forms.FormSubmit, {
        ref: ref
    }, t('general:submit'));
};
