// payload/dist/collections/operations/local/delete.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _errors = require("../../../errors");
const _setRequestContext = require("../../../express/setRequestContext");
const _init = require("../../../translations/init");
const _dataloader = require("../../dataloader");
const _delete = /*#__PURE__*/ _interop_require_default(require("../delete"));
const _deleteByID = /*#__PURE__*/ _interop_require_default(require("../deleteByID"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function deleteLocal(payload, options) {
    const { id, collection: collectionSlug, context, depth, fallbackLocale, locale = null, overrideAccess = true, req: incomingReq, showHiddenFields, user, where } = options;
    const collection = payload.collections[collectionSlug];
    const defaultLocale = payload?.config?.localization ? payload?.config?.localization?.defaultLocale : null;
    if (!collection) {
        throw new _errors.APIError(`The collection with slug ${String(collectionSlug)} can't be found. Delete Operation.`);
    }
    const req = {
        fallbackLocale: typeof fallbackLocale !== 'undefined' ? fallbackLocale : defaultLocale,
        i18n: (0, _init.i18nInit)(payload.config.i18n),
        locale: locale ?? defaultLocale,
        payload,
        payloadAPI: 'local',
        transactionID: incomingReq?.transactionID,
        user
    };
    (0, _setRequestContext.setRequestContext)(req, context);
    if (!req.t) req.t = req.i18n.t;
    if (!req.payloadDataLoader) req.payloadDataLoader = (0, _dataloader.getDataLoader)(req);
    const args = {
        id,
        collection,
        depth,
        overrideAccess,
        req,
        showHiddenFields,
        where
    };
    if (options.id) {
        return (0, _deleteByID.default)(args);
    }
    return (0, _delete.default)(args);
}
const _default = deleteLocal;
