// payload/dist/admin/components/views/Versions/columns.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "buildVersionColumns", {
    enumerable: true,
    get: function() {
        return buildVersionColumns;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
const _reacti18next = require("react-i18next");
const _reactrouterdom = require("react-router-dom");
const _ = require("../..");
const _formatDate = require("../../../utilities/formatDate");
const _SortColumn = /*#__PURE__*/ _interop_require_default(require("../../elements/SortColumn"));
const _Config = require("../../utilities/Config");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const CreatedAtCell = ({ id, collection, date, global })=>{
    const { admin: { dateFormat }, routes: { admin } } = (0, _Config.useConfig)();
    const { params: { id: docID } } = (0, _reactrouterdom.useRouteMatch)();
    const { i18n } = (0, _reacti18next.useTranslation)();
    let to;
    if (collection) to = `${admin}/collections/${collection.slug}/${docID}/versions/${id}`;
    if (global) to = `${admin}/globals/${global.slug}/versions/${id}`;
    return /*#__PURE__*/ _react.default.createElement(_reactrouterdom.Link, {
        to: to
    }, date && (0, _formatDate.formatDate)(date, dateFormat, i18n?.language));
};
const TextCell = ({ children })=>/*#__PURE__*/ _react.default.createElement("span", null, children);
const buildVersionColumns = (collection, global, t)=>[
        {
            name: '',
            accessor: 'updatedAt',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    label: t('general:updatedAt'),
                    name: "updatedAt"
                }),
                renderCell: (row, data)=>/*#__PURE__*/ _react.default.createElement(CreatedAtCell, {
                        collection: collection,
                        date: data,
                        global: global,
                        id: row?.id
                    })
            },
            label: ''
        },
        {
            name: '',
            accessor: 'id',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    disable: true,
                    label: t('versionID'),
                    name: "id"
                }),
                renderCell: (row, data)=>/*#__PURE__*/ _react.default.createElement(TextCell, null, data)
            },
            label: ''
        },
        {
            name: '',
            accessor: 'autosave',
            active: true,
            components: {
                Heading: /*#__PURE__*/ _react.default.createElement(_SortColumn.default, {
                    disable: true,
                    label: t('type'),
                    name: "autosave"
                }),
                renderCell: (row)=>/*#__PURE__*/ _react.default.createElement(TextCell, null, row?.autosave && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_.Pill, null, t('autosave')), "\xa0\xa0"), row?.version._status === 'published' && /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement(_.Pill, {
                        pillStyle: "success"
                    }, t('published')), "\xa0\xa0"), row?.version._status === 'draft' && /*#__PURE__*/ _react.default.createElement(_.Pill, null, t('draft')))
            },
            label: ''
        }
    ];
