// payload/dist/admin/components/graphics/File/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_default(require("react"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const File = ()=>/*#__PURE__*/ _react.default.createElement("svg", {
        height: "150",
        viewBox: "0 0 150 150",
        width: "150",
        xmlns: "http://www.w3.org/2000/svg"
    }, /*#__PURE__*/ _react.default.createElement("rect", {
        fill: "#333333",
        height: "150",
        transform: "translate(0 0.5)",
        width: "150"
    }), /*#__PURE__*/ _react.default.createElement("path", {
        d: "M82.8876 50.5H55.5555V100.5H94.4444V61.9818H82.8876V50.5Z",
        fill: "white"
    }), /*#__PURE__*/ _react.default.createElement("path", {
        d: "M82.8876 61.9818H94.4444L82.8876 50.5V61.9818Z",
        fill: "#9A9A9A"
    }));
const _default = File;
