// @payloadcms/richtext-slate/dist/field/elements/upload/Button/index.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _elements = require("payload/components/elements");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _slatereact = require("slate-react");
const _Upload = /*#__PURE__*/ _interop_require_default(require("../../../icons/Upload"));
const _Button = /*#__PURE__*/ _interop_require_default(require("../../Button"));
const _EnabledRelationshipsCondition = require("../../EnabledRelationshipsCondition");
const _injectVoid = require("../../injectVoid");
require("./index.scss");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'upload-rich-text-button';
const insertUpload = (editor, { relationTo, value })=>{
    const text = {
        text: ' '
    };
    const upload = {
        children: [
            text
        ],
        relationTo,
        type: 'upload',
        value
    };
    (0, _injectVoid.injectVoidElement)(editor, upload);
    _slatereact.ReactEditor.focus(editor);
};
const UploadButton = ({ enabledCollectionSlugs })=>{
    const { t } = (0, _reacti18next.useTranslation)([
        'upload',
        'general'
    ]);
    const editor = (0, _slatereact.useSlate)();
    const [ListDrawer, ListDrawerToggler, { closeDrawer }] = (0, _elements.useListDrawer)({
        collectionSlugs: enabledCollectionSlugs,
        uploads: true
    });
    const onSelect = (0, _react.useCallback)(({ collectionConfig, docID })=>{
        insertUpload(editor, {
            relationTo: collectionConfig.slug,
            value: {
                id: docID
            }
        });
        closeDrawer();
    }, [
        editor,
        closeDrawer
    ]);
    return /*#__PURE__*/ _react.default.createElement(_react.Fragment, null, /*#__PURE__*/ _react.default.createElement(ListDrawerToggler, null, /*#__PURE__*/ _react.default.createElement(_Button.default, {
        className: baseClass,
        el: "div",
        format: "upload",
        onClick: ()=>{
        // do nothing
        },
        tooltip: t('fields:addUpload')
    }, /*#__PURE__*/ _react.default.createElement(_Upload.default, null))), /*#__PURE__*/ _react.default.createElement(ListDrawer, {
        onSelect: onSelect
    }));
};
const _default = (props)=>{
    return /*#__PURE__*/ _react.default.createElement(_EnabledRelationshipsCondition.EnabledRelationshipsCondition, {
        ...props,
        uploads: true
    }, /*#__PURE__*/ _react.default.createElement(UploadButton, props));
};
